### PR TITLE
feat: memoize paths to root

### PR DESCRIPTION
**Pull request by admons**
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/dep-graph/blob/master/.github/CONTRIBUTING.md) rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?
today we calculate the paths to root every call to pathsFromNodeToRoot, even that the graph is "immutable"
there's no reason not to memoize it.

this change is using the same memoization technique as graphToDepTree to deal with cycles

I compared locally the performance for a dense graph between vuln's current `cachedPkgPathsToRoot` (1) vs this change (2)
for 10 runs in a row, (1) showed an average of 7618 ms, vs (2) that showed 3436 ms (55% less)

looking at this graph of daily total time of `prepareAllVulnerablePathsInfos`, the time that can be saved is up to 73 minutes of calculation a day (on 20/8, 7,940 seconds * 0.55 = 4367 seconds = ±73 minutes)
image

### CODE REVIEW

File: src/core/dep-graph.ts [hunk 1]
```diff
@@ -266,14 +273,16 @@ class DepGraphImpl implements types.DepGraphInternal {
...
     // Sort dependencies by name@version string.
-    const sortFn =
-      (graph: types.DepGraphInternal) => (idA: string, idB: string) => {
-        const pkgA = graph.getNodePkg(idA);
-        const pkgB = graph.getNodePkg(idB);
-        return DepGraphImpl.getPkgId(pkgA).localeCompare(
-          DepGraphImpl.getPkgId(pkgB),
-        );
-      };
+    const sortFn = (graph: types.DepGraphInternal) => (
```
Comment (admons) [inline src/core/dep-graph.ts:276]: This section is just prettier change

File: src/core/dep-graph.ts [hunk 2]
```diff
@@ -317,30 +326,56 @@ class DepGraphImpl implements types.DepGraphInternal {
   private pathsFromNodeToRoot(
```
Comment (admons) [inline src/core/dep-graph.ts:326]: This is a private method that I verified no one is using this - https://github.com/search?q=org%3Asnyk+pathsFromNodeToRoot&type=code

File: src/legacy/memiozation.ts [hunk 3]
```diff
@@ -1,49 +1,48 @@
-import { DepTree } from './index';
```
Comment (admons) [inline src/legacy/memiozation.ts:1]: This file turned to be generic rather than DepGraph specific, no logic change

File: src/core/dep-graph.ts [hunk 4]
```diff
@@ -2,6 +2,12 @@ import * as _isEqual from 'lodash.isequal';
 import * as graphlib from '../graphlib';
 import * as types from './types';
 import { createFromJSON } from './create-from-json';
+import {
+  getMemoizedItem,
+  MemoizationMap,
+  memoize,
+} from '../legacy/memiozation';
```
Comment (gjvis) [inline src/core/dep-graph.ts:9]: I haven't looked at the rest of this PR, but this is an issue – something from `core` mustn't have a dependency on something in `legacy`. Moving the memoization module also gives you an opportunity to fix the typo in its name :D

Comment (admons) [reply src/core/dep-graph.ts]: Thanks, fixed the typo and moved it to another folder

File: src/core/dep-graph.ts [hunk 5]
```diff
@@ -2,6 +2,8 @@ import * as _isEqual from 'lodash.isequal';
 import * as graphlib from '../graphlib';
 import * as types from './types';
 import { createFromJSON } from './create-from-json';
+import { getMemoizedItem, MemoizationMap, memoize } from '../utils/memoization';
```
Comment (admons) [inline src/core/dep-graph.ts:5]: Move to core

File: src/core/traversal-cache.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+  public set(nodeId: NodeId, item: T, partitionedCycles?: PartitionedCycles) {
+    if (!partitionedCycles) {
+      this.map.set(nodeId, { item });
+      return;
+    }
+    const { cyclesStartWithThisNode, cyclesWithThisNode } = partitionedCycles;
+    if (cyclesStartWithThisNode?.length > 0) {
+      // add explaination
+      const cycleNodeIds = new Set<NodeId>(...cyclesStartWithThisNode);
```
Comment (darscan) [inline src/core/traversal-cache.ts:30]: Please don't spread here. We've run into stack overflows by doing that. Also, it's unnecessary as `cyclesStartWithThisNode` is an array.

```suggestion
      const cycleNodeIds = new Set<NodeId>(cyclesStartWithThisNode);
```
```suggestion
      const cycleNodeIds = new Set<NodeId>(...cyclesStartWithThisNode);
```

File: src/core/traversal-cache.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+      this.map.set(nodeId, { item, cycleNodeIds });
+    } else if (!cyclesWithThisNode?.length) {
+      // add explaination
```
Comment (darscan) [inline src/core/traversal-cache.ts:33]: Reminder to add comment

File: src/core/dep-graph.ts [hunk 8]
```diff
@@ -127,9 +130,11 @@ class DepGraphImpl implements types.DepGraphInternal {
   public pkgPathsToRoot(pkg: types.Pkg): types.PkgInfo[][] {
     const pathsToRoot: types.PkgInfo[][] = [];
     for (const nodeId of this.getPkgNodeIds(pkg)) {
-      const pathsFromNodeToRoot = this.pathsFromNodeToRoot(nodeId);
+      const [pathsFromNodeToRoot] = this.pathsFromNodeToRoot(nodeId);
       for (const path of pathsFromNodeToRoot) {
-        pathsToRoot.push(path);
+        // the reason for the .slice() is to create a new path array for every call.
```
Comment (darscan) [inline src/core/dep-graph.ts:135]: You've added an unrelated functional change without adding tests

Comment (darscan) [reply src/core/dep-graph.ts:135]: Actually that's probably not true. Previously "paths" would be generated from scratch.. and so external mutations wouldn't matter. Now that we're caching things we can't let external callers mutate the data in the cache.

**Comment by darscan** [comment]
As discovered in https://github.com/snyk/dep-graph/pull/105 memoizing paths-to-root when cycles are involved is *harder* than it seems. This PR is likely now stale, and I feel we should probably start from scratch if we were to pick this up again.

**Pull request closed without merging**